src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -852,29 +852,61 @@ Shard::finalizeInternal(
         if (detShard)
             detShard->store(nodeObject);
 
-        std::lock_guard lock(mutex_);
-
-        // Remove the acquire SQLite database
-        if (acquireInfo_)
         {
-            acquireInfo_.reset();
-            remove_all(dir_ / AcquireShardDBName);
-        }
+            std::lock_guard lock(mutex_);
 
-        if (!initSQLite(lock))
-            return fail("failed to initialize SQLite databases");
+            // Remove the acquire SQLite database
+            if (acquireInfo_)
+            {
+                acquireInfo_.reset();
+                remove_all(dir_ / AcquireShardDBName);
+            }
 
-        setFileStats(lock);
-        lastAccess_ = std::chrono::steady_clock::now();
+            if (!initSQLite(lock))
+                return fail("failed to initialize SQLite databases");
+
+            setFileStats(lock);
+            lastAccess_ = std::chrono::steady_clock::now();
+        }
 
         if (detShard)
         {
-            /* Close non-deterministic shard database. */
+            // Do not allow all other threads work with the shard
+            busy_ = true;
+
+            // Wait until all other threads leave the shard
+            while (backendCount_ > 1)
+                std::this_thread::yield();
+
+            // Obtain the mutex
+            std::lock_guard lock(mutex_);
+
+            // Close original shard database
             backend_->close();
-            /* Replace non-deterministic shard by deterministic one. */
+
+            // Close SQL databases for original shard
+            lgrSQLiteDB_.reset();
+            txSQLiteDB_.reset();
+            acquireInfo_.reset();
```

Comment:
No need to reset `acquireInfo_` again, we just reset it at line 861.

Reply:
Done.